### ISSUE TITLE
Support a large query with facet, or hi-light feature. [JIRA: RIAK-1875]

### ISSUE THREAD

**Issue by ksauzz** [basho/yokozuna#503]
#### Problem:

Riak Search has a limitation of the max size of a search query if a user uses hi-light or facet feature.

In HTTP API, there is the max http header size (default is 8192*2 on LInux) due to hard-coded recbuf in mochiweb. [[1](https://github.com/basho/mochiweb/blob/ade2a9b29a11034eb550c1d79b4f991bf5ca05ba/include/internal.hrl#L2)]   In Protocol Buffer API, as message representations doesn't support hi-light or facet yet, a user can't use facet or hi-light feature on any PB clients. [[2](https://github.com/basho/riak_pb/blob/620bc7001dc788e5530078aa8be53c9d15d4fdb4/src/riak_search.proto#L38-L55)] There is no workaround to use a query over the max size with the features. A customer said their end-users often use a such large query because of multibyte character set, and complicated condition in a query. see: zd#11091

Note: [recbuf's behaviour depends on a platform.](https://github.com/mochi/mochiweb/blob/b66b68d95c9e1b2a32b194202e870e6d1e232eb7/src/mochiweb_socket_server.erl#L122-L133)
#### Solutions:
1. Expose recbuf option: As the latest mochi/mochiweb has the change of exposing recbuf, we can expose the option by applying new changes in mochi/mochiweb, and updating webmachine a bit. It's easier than the solution 2. see: https://github.com/basho/mochiweb/issues/18
2. Support 'POST' method for search query: Querying by post method is not affected by the limitation of recbuf. 

---
##### Error response when hitting the limitation of the request header size in HTTP API.

```
% curl "localhost:8098/search/query/my_index?q=`for i in $(seq 1 20000);do echo -n 1;done`" -i
HTTP/1.1 400 Bad Request
Server: MochiWeb/1.0 (Any of you quaids got a smint?)
Date: Fri, 19 Jun 2015 05:33:25 GMT
Content-Length: 0
```

**Comment by Basho-JIRA** [comment]
@doug I added 2.0.7 to fix version.  I can see in GH it was merged to 2.0 branch.  I didn't add 2.2 though because I don't know if it landed there.  only see one PR.  

**_[posted via JIRA by Patricia Brewer]**_

**Comment by Basho-JIRA** [comment]
It was actually released in Riak 2.1.2, so should already be in 2.2.0 once we get there. 

**_[posted via JIRA by Douglas Rohrer]**_
